### ISSUE TITLE
default seccomp profile is slow (on aarch64?)

### ISSUE THREAD

**Issue by martinetd** [containers/podman#13467]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug ?

**Description**

I was pointed out by one of our users that some benchmark workload ran much slower within a container than without (about half speed).
(For reference, it was https://github.com/kdlucas/byte-unixbench 's "pipe" workload and another one there, pipe is a simple loop that writes/read a pipe so for all intent and purposes I will illustrate below with a simple dd command)

Looking at it with perf record/report it turns out about half the time is spend inside seccomp, within the bpf program used to determine whether to allow the call or not. Here is the write side:
```
 --39.81%--el0_svc_common.constprop.0
           |
           |--18.87%--__arm64_sys_write
           |          |
           |           --18.47%--ksys_write
           |                     |
           |                      --17.05%--vfs_write
           |                                |
           |                                |--14.22%--new_sync_write
           |                                |          |
           |                                |          |--11.29%--pipe_write
           |                                |          |          |
           |                                |          |          |--3.31%--copy_page_from_iter
           |                                |          |          |          |
           |                                |          |          |           --2.32%--__arch_copy_from_user
           |                                |          |          |
           |                                |          |          |--2.63%--__wake_up_sync_key
           |                                |          |          |          |
           |                                |          |          |          |--1.74%--_raw_spin_unlock_irqrestore
           |                                |          |          |          |
           |                                |          |          |           --0.55%--__wake_up_common_lock
           |                                |          |          |
           |                                |          |           --1.70%--file_update_time
           |                                |          |                     |
           |                                |          |                      --0.57%--current_time
           |                                |          |
           |                                |           --0.70%--_raw_spin_unlock_irq
           |                                |
           |                                 --0.62%--fsnotify
           |
            --17.50%--syscall_trace_enter
                      |
                       --16.87%--__secure_computing
                                 |
                                  --16.57%--__seccomp_filter
                                            |
                                             --15.25%--__bpf_prog_run32
                                                       |
                                                        --15.00%--___bpf_prog_run
```

It turns out that we generate a program that is quite linear: for each syscall listed we painstakingly check it and move on to the next, so having a large seccomp profile gets slower as it grows, see next section.

It's possible to do much better now with constant-action bitmaps:
https://lwn.net/Articles/834785/

**Steps to reproduce the issue:**

```
# for profile in unconfined *json; do
      echo -n "$profile: "; 
      podman run -ti --rm --security-opt=seccomp=$profile docker.io/alpine \
              time dd if=/dev/zero of=/dev/null bs=512 count=1000000 status=none \
               | awk '/real/ { print $3 }';
      printf "\r";
  done
unconfined: 1.21s
onlydef.json: 1.73s
simpler.json: 1.82s 
simpler-fluff.json: 2.30s
seccomp.json: 3.00s
```
Where:
- unconfined is no seccomp
- onlydef is a profile with only default set and no syscall listed: https://gaia.codewreck.org/local/tmp/podman/onlydef.json
- simpler is a profile that defaults to deny with only strictly needed syscalls listed: https://gaia.codewreck.org/local/tmp/podman/simpler.json
- simpler-fluff is the same profile with some more syscalls listed just to show that adding more (never called) syscalls slows it down. I've also tried moving read/writes top and bottom of list, but that doesn't impact the performance, I assume the list is reordered somewhere when generating the program. https://gaia.codewreck.org/local/tmp/podman/simpler-fluff.json
- seccomp.json is the default profile I found on my system (podman 3.4.4) https://gaia.codewreck.org/local/tmp/podman/seccomp.json

**Describe the results you received:**

Stuff is slow.

**Describe the results you expected:**

Stuff should be fast! (e.g. shouldn't slow down when adding more syscalls)

**Additional information you deem important (e.g. issue happens only occasionally):**

It doesn't seem to be as bad on my laptop (x86_64), I didn't check why.
I'll also admit I don't understand *what* generates the bpf program from the json profile, so while I had a quick look at other issues various git logs to see if it would be fixed upstreams, I'm not 100% sure it isn't. For all I know it might be the responsibility of libseccomp in which case I'm barking at the wrong tree... Please advise.

Thanks!

**Comment by baude** [comment]
@giuseppe do you know if this should be simply filed with seccomp folks ?  I wouldn't think our seccomp.json is the culprit?

**Comment by mheon** [comment]
Quite possible it is. Our profile is *gigantic* - I think we have every syscall in the kernel in there due to the errno split for default action. Means every single syscall has to potentially traverse all 300+ rules to be allowed/denied. This is significantly larger than Docker's, for example.

The libseccomp code could certainly be made better at structuring the rules such that it's not O(n) to find a match, but for now, we could investigate making frequently-used syscalls higher priority in the rules, and thus more likely to be hit early, not requiring a long traversal.

**Comment by giuseppe** [comment]
Constant actions is an internal optimization in the kernel, that is done transparently.  There is nothing to do from userspace.  Could you please check yhat your kernel is new enough that the feature is there?

**Comment by martinetd** [comment]
> Constant actions is an internal optimization in the kernel, that is done transparently. There is nothing to do from userspace. Could you please check yhat your kernel is new enough that the feature is there?

Ugh, sorry -- the patch I linked at was never merged, and instead f9d480b6ffbe ("seccomp/cache: Lookup syscall allowlist bitmap for fast path") got in which apparently isn't in my kernel (merged in 5.11, and am on 5.10 and there is no backport), so I guess that's it.

I don't get it though -- it's userspace's code that compiles the bpf program, and the optimization done in kernel could very well be done in that generated program, why did they have to go out of their way to guess how the program works and provide a cache.. sometimes.. if stars are aligned... (in particular more complex profiles that depend on syscall arguments can't be cached like this, so not everyone benefit from it) :|

Oh, well. Sorry for the noise. I'll reopen after checking if perfs are still as slow after picking these patches...
